src/csharp/Grpc.Microbenchmarks/CommonThreadedBase.cs

```diff
@@ -0,0 +1,66 @@
+#region Copyright notice and license
+
+// Copyright 2015 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Grpc.Core;
+
+namespace Grpc.Microbenchmarks
+{
+
+    // common base-type for tests that need to run with some level of concurrency;
+    // note there's nothing *special* about this type - it is just to save some
+    // boilerplate
+
+    [ClrJob, CoreJob] // test .NET Core and .NET Framework
+    [MemoryDiagnoser] // allocations
+    public abstract class CommonThreadedBase
+    {
+        protected virtual bool NeedsEnvironment => true;
+
+        [Params(1, 1, 2, 4, 8, 12)]
+        public int ThreadCount { get; set; }
+
+        protected GrpcEnvironment Environment { get; private set; }
+
+        [GlobalSetup]
+        public virtual void Setup()
+        {
+            ThreadPool.GetMinThreads(out var workers, out var iocp);
+            if (workers <= ThreadCount) ThreadPool.SetMinThreads(ThreadCount + 1, iocp);
+            if (NeedsEnvironment) Environment = GrpcEnvironment.AddRef();
+        }
+
+        [GlobalCleanup]
+        public virtual void Cleanup()
+        {
+            if (Environment != null)
+            {
+                Environment = null;
+                GrpcEnvironment.ReleaseAsync().Wait();
+            }
+        }
+
+        protected void RunConcurrent(Action operation)
+        {
+            Parallel.For(0, ThreadCount, _ => operation());
```

Comment:
I'm concerned if BenchmarkDotNet will give good results for multi-threaded microbenchmarks, seems that it hasn't been designed for that job.
E.g. here: https://github.com/dotnet/BenchmarkDotNet/issues/5

Reply:
My understanding here is that benchmarkdotnet doesn't have any *specific features* for multi-threaded tests, and yes it isn't it's strong suit; but I don't think that it is any better or worse than doing the same thing manually with a hand-written rig. But with benchmarkdotnet you can still get a lot more additional data / features etc for free, despite it not being the perfect scenario.

What *is* for sure is that global allocations (all threads) is missing an API on .NET Core < v3, but that this new API is already hooked up in the v3 bits including benchmarkdotnet - so when v3 lands, I strongly recommend updating any benchmarks to netcoreapp3.0 (either additional-to or instead-of) - this will make the "Allocated" column more reliable.

For my money: we only gain by switching to benchmarkdotnet; but: your call.

Reply:
I agree that by switching to benchmarkDotNet we will only gain. I'm just trying to understand what drawbacks are there regarding to the multi-threaded use.
I'm currently experimenting with `OperationsPerInvoke = 1000000` to make sure the overhead of Parallel.For and the thread jumps etc are negligible compared to the workload.

Reply:
since the async operation is sync, not async, there won't be massive numbers of thread jumps involved here; the main risk is that some of the operations end up getting queued rather than running in parallel; I've tried making sure that we have at least as many workers as we need, but indeed - the cost of spinning up an entire thread *in the `[Benchmark]` method isn't really viable*; that said, because of the warmup etc, we should be fairly happy that the thread-pool will have adapted to our usage reasonably. While thread-pool growth is pretty sedate by default, once they are there: they *really, really* don't like to terminate themselves.